bpf/bpf_host.c

```diff
@@ -652,21 +730,63 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 }
 
 static __always_inline int
-tail_handle_ipv4(struct __ctx_buff *ctx, __u32 ipcache_srcid, const bool from_host)
+tail_handle_ipv4_cont(struct __ctx_buff *ctx, bool from_host)
 {
 	__u32 proxy_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
 	int ret;
 	__s8 ext_err = 0;
 
 	ctx_store_meta(ctx, CB_SRC_LABEL, 0);
 
-	ret = handle_ipv4(ctx, proxy_identity, ipcache_srcid, from_host, &ext_err);
+	ret = handle_ipv4_cont(ctx, proxy_identity, from_host, &ext_err);
 	if (IS_ERR(ret))
 		return send_drop_notify_error_ext(ctx, proxy_identity, ret, ext_err,
 						  CTX_ACT_DROP, METRIC_INGRESS);
 	return ret;
 }
 
+declare_tailcall_if(is_defined(ENABLE_HOST_FIREWALL), CILIUM_CALL_IPV4_CONT_FROM_HOST)
+int tail_handle_ipv4_cont_from_host(struct __ctx_buff *ctx)
+{
+	return tail_handle_ipv4_cont(ctx, true);
+}
+
+declare_tailcall_if(is_defined(ENABLE_HOST_FIREWALL), CILIUM_CALL_IPV4_CONT_FROM_NETDEV)
+int tail_handle_ipv4_cont_from_netdev(struct __ctx_buff *ctx)
+{
+	return tail_handle_ipv4_cont(ctx, false);
+}
+
+static __always_inline int
+tail_handle_ipv4(struct __ctx_buff *ctx, __u32 ipcache_srcid, const bool from_host)
+{
+	__u32 proxy_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
+	int ret;
+
+	ctx_store_meta(ctx, CB_SRC_LABEL, 0);
+
+	ret = handle_ipv4(ctx, proxy_identity, ipcache_srcid, from_host);
+
+	/* TC_ACT_REDIRECT is not an error, but it means we should stop here. */
+	if (ret == CTX_ACT_OK) {
+		if (from_host)
+			invoke_tailcall_if(is_defined(ENABLE_HOST_FIREWALL),
+					   CILIUM_CALL_IPV4_CONT_FROM_HOST,
```

Comment:
Probably just a matter of taste. But I would prefer to have the `ctx_store_meta(ctx, CB_SRC_LABEL, secctx)` here, and not hidden inside `handle_ipv4()`. It's already difficult enough to keep track of all the games we play with `CB_SRC_LABEL` ...

Reply:
`CB_FROM_HOST` should probably stay in it current place - as that's where we create its input data.

I'm just very sensitive to `CB_SRC_LABEL`, burned too many times already :smile:.